### DISCUSSION TITLE
hyperlinks look the same as text

### DISCUSSION THREAD

**Discussion by flying-strawberry** [ghostty-org/ghostty#2810] [Q&A]
Ghostty has support for terminal hyperlinks but is there a way to distinguish them from other text.
```
printf '\e]8;;https://example.com\e\\This is a link\e]8;;\e\\\n'
```

**Comment by 00-kat** [comment]
I should mention, holding down <kbd>ctrl</kbd> (Linux) or <kbd>cmd</kbd> (macOS) and hovering over a link will underline it.

I do believe that there should be some indication without doing this, so I'm keeping this discussion open as “indication without holding down modifier” (which may have actually been what was actually requested).  There was some related previous discussion in #3903 and #4382.

**Reply by PerBothner** [reply]
One suggestion: For OSC 8 hyperlinks, default the text foreground-color to blue, if no other color is specified. This matches the traditional default display of links in web browsers.  Ideally, this could be configured and/or theme-dependent.

**Comment by kaddkaka** [comment]
Requiring a modifier keyy makes it very unlikely to discover a hyperlink unless the visible text looks like a uri in itself. 

Static decorator like
- background color change
- underline, or
- first character foreground color  
even when not hovering the link might be ok, but might also be too intrusive. (are there escape codes to print text with underline that the tui application programmer can use to indicate hyperlinks?) 

Having an indication simply on hover would greatly improve discoverability. I would also prefer to not have to hold ctrl for a click to follow the link.

**Reply by rhodes-b** [reply]
Its very expensive (slow) to parse all the text then decide if its a link. This is why ghostty requires a modifier and only works under the cursor

Since ghostty doesnt support hints (yet). You should only be looking for a uri which you said is discoverable on its own.

**Reply by 00-kat** [reply]
> are there escape codes to print text with underline that the tui application programmer can use to indicate hyperlinks?

Not certain what you mean here, but:
- SGR 4 can underline text: `printf '\033[4mHello, world!\033[0m\n'`.
- Curly underlines and a few other types are possible with subparameters to SGR 4, on terminal emulators which implement [Kitty's extension] for it (which includes Ghostty).  See that page for more details.
- Underline colour is also possible with [Kitty's extension] using SGR 59.
- Text can be marked as an underline using [VTE+iTerm2's OSC 8]: `printf '\033]8;;https://example.com\033\\Hello, example.com\033]8;;\033\\!\n'`.  One of the VTE maintainers also has [a Gist about this].

[Kitty's extension]: https://sw.kovidgoyal.net/kitty/underlines/
[VTE+iTerm2's OSC 8]: https://iterm2.com/3.2/documentation-escape-codes.html#:~:text=Anchor
[a Gist about this]: https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda

**Reply by 00-kat** [reply]
> Its very expensive (slow) to parse all the text then decide if its a link.

Would it not be possible to only parse the text when the cursor moves over it?  Or parse text when it is written to the screen and mark it as an underline?

**Reply by kaddkaka** [reply]
I don't really understand that. The fact that it's a link is already known as soon as the text has been rendered. 

When moving the mouse, just see if it is over one of those known links (no parsing needed to trigger when moving the mouse). 

Or is it that all mouse movement is costly to report?

**Reply by 00-kat** [reply]
> The fact that it's a link is already known as soon as the text has been rendered.

Ghostty supports both using OSC 8 and has a regular expression to detect whether some text looks like a hyperlink; the latter is what is expensive.  Perhaps this feature could be restricted to only OSC 8, since the user can already tell if a textual hyperlink is a hyperlink just by looking at it

**Reply by pluiedev** [reply]
I personally think it makes sense to restrict this to OSC 8, in which case it's guaranteed that the intent is to render the text as a hyperlink. Having to do guesswork to parse a link opens us up to false positives, which is a worse outcome IMO. What is less clear to me is what styling we should apply to the link, and how the default styling should interact with any other style overrides.

For now I think a simple underline would be more than okay, which can then be overridden by CSIs. I don't really think we should make it configurable since that introduces a needlessly complex configuration interface, but if people make a compelling enough case for it then... who knows?

**Reply by kaddkaka** [reply]
Suggestion for OSC 8:

Don't affect how the unhovered link text is displayed. 

Toggle the underline style when the hyperlink text is hovered (and change mouse cursor)

**Reply by kaddkaka** [reply]
> I personally think it makes sense to restrict this to OSC 8, in which case it's guaranteed that the intent is to render the text as a hyperlink. Having to do guesswork to parse a link opens us up to false positives, which is a worse outcome IMO.

False positives would be annoying, but are they a big risk if the match has to start with `http://` or `https://`? I'm not familiar with the details of URLs, is it impossible to reliably tell where the end of the URL is? 

Also, when outputting text, would it be more computational heavy to detect http than detecting the OSC codes?

**Reply by pluiedev** [reply]
For reference, this is how mindbogglingly complex our URL regex is: https://github.com/ghostty-org/ghostty/blob/main/src/config/url.zig#L25-L33

URLs aren't at all trivial to parse, and what you are proposing isn't feasible without sacrificing performance.

**Reply by 00-kat** [reply]
> our URL regex

The comment above it also notes that this fails for complex cases.

**Reply by kaddkaka** [reply]
> For reference, this is how mindbogglingly complex our URL regex is: https://github.com/ghostty-org/ghostty/blob/main/src/config/url.zig#L25-L33
> 
> URLs aren't at all trivial to parse, and what you are proposing isn't feasible without sacrificing performance.

In that case I agree, doing it for only OSC8 sounds good.

How simple is OSC8 to detect?

**Reply by pluiedev** [reply]
We already keep track of them — the only thing that needs to be done here is to render them differently

**Reply by tiesmaster** [reply]
I've also ran into the issue that URLs are pretty difficult to detect, and found this discussion. It looks like the consensus is to make URLs better discoverable for OSC8, shall I make an issue for this?

**Reply by rhodes-b** [reply]
No issues are only made by maintainers once its a accepted feature request

**Reply by michaelfortunato** [reply]
Interested in this as well--having an option to have hover automatically underline the link (no modifier needed). Not sure how complex a PR would be. I am hoping the hover detection code could be reused from whatever code was written for the modifier case.

**Comment by danez** [comment]
I like how xterm.js in vscode does it:

<img width="283" height="104" alt="Screenshot 2026-04-03 at 01 45 25" src="https://github.com/user-attachments/assets/8ec109f6-ae62-4106-b2ca-f4a76ab0a0a1" />

I underlines with dashed lines any hyperlink.

**Reply by j4james** [reply]
FWIW, in Gnome Terminal (which is basically the reference implementation for `OSC 8` hyperlinks), they display the linked text with a dotted underline by default, which changes to a solid underline when hovered. Actually activating the link requires a Ctrl+Click.

**Comment by landsman** [comment]
I really need this!

**Reply by landsman** [reply]
I would love to fix that via my own script or configuration, so related to: https://github.com/ghostty-org/ghostty/discussions/2353
